### ISSUE TITLE
Bug: time slice not work in react 18

### ISSUE THREAD

**Issue by yuwanlin** [facebook/react#24392]
in react@16.8.0, a long task will be sliced multi short task, demo: https://stackblitz.com/edit/react-ts-aqwejz
![image](https://user-images.githubusercontent.com/18443683/163744795-3abfb632-961f-4ea7-a669-f37e65d59f4c.png)

but in react@18.0.0, it will be only a long task, demo: https://stackblitz.com/edit/react-ts-ezgtzn
![image](https://user-images.githubusercontent.com/18443683/163745038-dbf8ba83-8988-453b-b7cb-a217abe29db1.png)

Is it a react18 time slice bug or feature? 

React version: 16.8.0 & 18.0.0

## Steps To Reproduce

1. run a example app like below
2. open inspector -> performance, then record and analyze

Link to code example: 
16.8.0: https://stackblitz.com/edit/react-ts-aqwejz
18.0.0: https://stackblitz.com/edit/react-ts-ezgtzn

## The current behavior
in react@18.0.0, long task not be sliced

## The expected behavior
in react@18.0.0, long task will be sliced

**Comment by gaearon** [comment]
Interesting. I thought adding startTransition would make it work, but it doesn't: https://stackblitz.com/edit/react-ts-f7zkug?file=index.tsx

We'll have a look.

**Comment by yuwanlin** [comment]
In my attempt, use React.startTransition make it work, see video record. In my opinion, default update lane is SyncLane, so it has the Highest priority, and is uninterruptible?



https://user-images.githubusercontent.com/18443683/163906889-eac541a3-9514-47e2-9159-625ff24bcd2e.mp4

.

**Comment by lishion** [comment]
hi [gaearon](https://github.com/gaearon), did you mean the time slice will not enable defaule unless create task in **startTransition**?

**Comment by lishion** [comment]
i think this is becase these is a condition to enable time slice in v18:
```js
const shouldTimeSlice =
    !includesBlockingLane(root, lanes) &&
    !includesExpiredLane(root, lanes) &&
    (disableSchedulerTimeoutInWorkLoop || !didTimeout);
```
And in includesBlockingLane function:
```
export function includesBlockingLane(root: FiberRoot, lanes: Lanes) {
 // will not hit this condition by default
  if (
    allowConcurrentByDefault &&
    (root.current.mode & ConcurrentUpdatesByDefaultMode) !== NoMode
  ) {
    // Concurrent updates by default always use time slicing.
    return false;
  }

  const SyncDefaultLanes =
    InputContinuousHydrationLane |
    InputContinuousLane |
    DefaultHydrationLane |
    DefaultLane;
  return (lanes & SyncDefaultLanes) !== NoLanes; // here only task created in startTransition can retrun false
}
```
So now, only the task created in startTransaction will enable the time slice. I'm not sure whether this is a bug or by design.

**Comment by yuwanlin** [comment]
@lishion I think you are right. enable time slice by `shouldTimeSlice `.  In [official website of v18's blog](https://zh-hans.reactjs.org/blog/2022/03/29/react-v18.html), It has a paragraph describes:
>A key property of Concurrent React is that rendering is interruptible. When you first upgrade to React 18, before adding any concurrent features, updates are rendered the same as in previous versions of React — in a single, uninterrupted, synchronous transaction. With synchronous rendering, once an update starts rendering, nothing can interrupt it until the user can see the result on screen.

**Comment by yuwanlin** [comment]
the key point is variable [shouldTimeSlice](https://github.com/facebook/react/blob/7cf8dfd94f29a42102e5e553395d9bb1b1710c42/packages/react-reconciler/src/ReactFiberWorkLoop.new.js#L898), when it's value is true, render will be sliced, or render will be continuous. 
```
const shouldTimeSlice =
    !includesBlockingLane(root, lanes) &&
    !includesExpiredLane(root, lanes) &&
    (disableSchedulerTimeoutInWorkLoop || !didTimeout);

let exitStatus = shouldTimeSlice
    ? renderRootConcurrent(root, lanes)
    : renderRootSync(root, lanes);
```
`includesBlockingLane` is below: 
```
export function includesBlockingLane(root: FiberRoot, lanes: Lanes) {
  const SyncDefaultLanes =
    InputContinuousHydrationLane |
    InputContinuousLane |
    DefaultHydrationLane |
    DefaultLane;
  return (lanes & SyncDefaultLanes) !== NoLanes;
}
```
in react 18, when we render App directly, the [lane of root](https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberReconciler.new.js#L332) is DefaultLane, so `includesBlockingLane` return true. `shouldTimeSlice` is false, then will run into `renderRootSync`.
```
const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(<App />);
```
when we render App use startTransition like below, the [lane of root](https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberReconciler.new.js#L332) is TransitionLane1. so `includesBlockingLane` return false. `shouldTimeSlice` is true, then will run into `renderRootConcurrent`.
```
const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement); 
React.startTransition(() => {
    root.render(<App />);
});
```
the main deference between `renderRootConcurrent` with `renderRootSync` is using deferent workLoop in their inner.  

`renderRootConcurrent` use `workLoopConcurrent`
```
function workLoopConcurrent() {
  // Perform work until Scheduler asks us to yield
  while (workInProgress !== null && !shouldYield()) {
    performUnitOfWork(workInProgress);
  }
}
```
`renderRootSync` use `workLoopSync`
```
function workLoopSync() {
  // Already timed out, so perform work without checking if we need to yield.
  while (workInProgress !== null) {
    performUnitOfWork(workInProgress);
  }
}
```
`workLoopConcurrent`'s condition has `shouldYield` :
```
function shouldYieldToHost() {
  var timeElapsed = exports.unstable_now() - startTime;
  if (timeElapsed < frameInterval) {
    // The main thread has only been blocked for a really short amount of time;
    // smaller than a single frame. Don't yield yet.
    return false;
  } // The main thread has been blocked for a non-negligible amount of time. We


  return true;
}
```
and `frameInterval`'s value is 5ms, so if we are in TransitionLanes, we will see many task and every task will cost about 5ms.
![image](https://user-images.githubusercontent.com/18443683/175461811-58ae634a-f8d0-4cce-89a2-fc15c564895d.png)

**Comment by monkindey** [comment]
hi, Any Update on it?

**Comment by rickhanlonii** [comment]
In older experimental versions, `createRoot` as time sliced by default. During the rollout of React 18, we switched this so that `createRoot` is sync by default, and you opt-into time-slicing with a concurrent feature like `startTransition` or Suspense. The `shouldTimeSlice` check mentioned above is where we do that. 

I just checked @gaearon's code, and I see that it is time-sliced when using `startTransition`:


```js
const root = createRoot(document.getElementById('root'));

React.startTransition(() => {
  root.render(<App />);
});
```
<img width="1506" alt="Screen Shot 2022-09-05 at 10 57 26 PM" src="https://user-images.githubusercontent.com/2440089/188537673-88fa6727-3cfc-4b81-a811-dcfcf138af2f.png">
